website/docs/plugin/framework/handling-data/dynamic-data.mdx

```diff
@@ -0,0 +1,167 @@
+    case types.List:
+        // Handle list value
+    case types.Map:
+        // Handle map value
+    case types.Object:
+        // Handle object value
+    case types.Set:
+        // Handle set value
+    case types.String:
+        // Handle string value
+    case types.Tuple:
+        // Handle tuple value
+    }
+```
+
+When writing test configurations and debugging provider issues, developers will also want to understand how Terraform represents [complex type literals](/terraform/language/expressions/type-constraints#complex-type-literals). For example, Terraform does not provide any way to directly represent lists, maps, or sets.
+
+### Understanding Type Consistency
+
+For [managed resources](/terraform/plugin/framework/resources), Terraform core implements [data consistency rules](https://github.com/hashicorp/terraform/blob/main/docs/resource-instance-change-lifecycle.md) between configuration, plan, and state data. With [dynamic attributes](/terraform/plugin/framework/handling-data/attributes/dynamic), these consistency rules are also applied to the **type** of data.
+
+For example, given a dynamic `example_attribute` that is computed and optional:
+
+```go
+func (r ThingResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+    resp.Schema = schema.Schema{
+        Attributes: map[string]schema.Attribute{
+            "example_attribute": schema.DynamicAttribute{
+                Computed: true,
+                Optional: true,
+                // ... potentially other fields ...
+            },
+            // ... potentially other attributes ...
+        },
+    }
+}
+```
+
+If a practitioner configures this resource as:
+
+```hcl
+resource "examplecloud_thing" "example" {
+  # This literal expression is a tuple[string, string]
+  example_attribute = ["one", "two"]
+}
+```
+
+Then the exact type must be planned and stored in state during `apply` as a [tuple](/terraform/plugin/framework/handling-data/types/tuple) with two [string](/terraform/plugin/framework/handling-data/types/string) element types. If provider code attempts to store this attribute as a different type, like a [list](/terraform/plugin/framework/handling-data/types/list) of strings, even with the same data values, Terraform will produce an error during apply:
+
+```bash
+│ Error: Provider produced inconsistent result after apply
+│ 
+│ When applying changes to examplecloud_thing.example, provider "provider[\"TYPE\"]" produced an unexpected new value: .example_attribute: wrong final value type: tuple required.
+│ 
+│ This is a bug in the provider, which should be reported in the providers own issue tracker.
+```
+
+If a practitioner configures this same resource as:
+
+```hcl
+resource "examplecloud_thing" "example" {
+  example_attribute = tolist(["one", "two"])
+}
+```
+
+Then the exact type must be planned and stored in state during `apply` as a [list](/terraform/plugin/framework/handling-data/types/list) of strings.
```

Comment:
Perhaps expand this section to say a little more, or provide examples, about the handling/behaviour of Terraform list/tuple, map/object, and framework list, map, object and set types in this regard?